### PR TITLE
GH-40153: [Python] Make `Tensor.__getbuffer__` work on 32-bit platforms

### PR DESCRIPTION
**Pull request by pitrou**
### Rationale for this change

`Tensor.__getbuffer__` would silently assume that `Py_ssize_t` is the same width as `int64_t`, which is true only on 64-bit platforms.

### What changes are included in this PR?

Create an internal buffer of `Py_ssize_t` values mirroring a Tensor's shape and strides, to avoid relying on the aforementioned assumption.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

No.

* GitHub Issue: #40153

### CODE REVIEW

**Comment by pitrou** [comment]
@github-actions crossbow submit -g python

File: python/pyarrow/tensor.pxi [hunk 1]
```diff
@@ -282,10 +292,8 @@ strides: {0.strides}""".format(self)
             buffer.readonly = 0
         else:
             buffer.readonly = 1
-        # NOTE: This assumes Py_ssize_t == int64_t, and that the shape
-        # and strides arrays lifetime is tied to the tensor's
-        buffer.shape = <Py_ssize_t *> &self.tp.shape()[0]
-        buffer.strides = <Py_ssize_t *> &self.tp.strides()[0]
+        buffer.shape = <Py_ssize_t *> cp.PyBytes_AsString(self._ssize_t_shape)
+        buffer.strides = <Py_ssize_t *> cp.PyBytes_AsString(self._ssize_t_strides)
```
[the remark this answers is not in the document]
Comment (pitrou) [reply python/pyarrow/tensor.pxi:296]: I don't think Cython would take the `bytes` object pointer implicitly? That said, I can try to remove it.

[the remark this answers is not in the document]
Comment (pitrou) [reply python/pyarrow/tensor.pxi:296]: I confirm it does not work:
```
  Error compiling Cython file:
  ------------------------------------------------------------
  ...
          buffer.obj = self
          if self.tp.is_mutable():
              buffer.readonly = 0
          else:
              buffer.readonly = 1
          buffer.shape = <Py_ssize_t *> self._ssize_t_shape
                         ^
  ------------------------------------------------------------

  pyarrow/tensor.pxi:295:23: Python objects cannot be cast to pointers of primitive types
```

and also:
```
  Error compiling Cython file:
  ------------------------------------------------------------
  ...
          buffer.obj = self
          if self.tp.is_mutable():
              buffer.readonly = 0
          else:
              buffer.readonly = 1
          buffer.shape = self._ssize_t_shape
                             ^
  ------------------------------------------------------------

  pyarrow/tensor.pxi:295:27: Cannot convert Python object to 'Py_ssize_t *'
```

**Comment by pitrou** [comment]
Any other comment @jorisvandenbossche ?

**Review by jorisvandenbossche — APPROVED** [review]

**Pull request merged**